eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs

```diff
@@ -274,21 +273,24 @@ ltePosInfty (SJust i) j = i <= j
 
 evalTimelock ::
   Era era =>
-  Set (KeyHash 'Witness (EraCrypto era)) ->
+  Set.Set (KeyHash 'Witness (EraCrypto era)) ->
   ValidityInterval ->
   Timelock era ->
   Bool
-evalTimelock _vhks (ValidityInterval txStart _) (RequireTimeStart lockStart) =
-  lockStart `lteNegInfty` txStart
-evalTimelock _vhks (ValidityInterval _ txExp) (RequireTimeExpire lockExp) =
-  txExp `ltePosInfty` lockExp
-evalTimelock vhks _vi (RequireSignature hash) = member hash vhks
-evalTimelock vhks vi (RequireAllOf xs) =
-  all (evalTimelock vhks vi) xs
-evalTimelock vhks vi (RequireAnyOf xs) =
-  any (evalTimelock vhks vi) xs
-evalTimelock vhks vi (RequireMOf m xs) =
-  m <= sum (fmap (\x -> if evalTimelock vhks vi x then 1 else 0) xs)
+evalTimelock vhks (ValidityInterval txStart txExp) = go
+  where
+    -- The important part of this validator is that it will stop as soon as it reaches the
+    -- required number of valid scripts
+    isValidMOf n SSeq.Empty = n <= 0
+    isValidMOf n (ts SSeq.:<| tss) =
+      n <= 0 || if go ts then isValidMOf (n - 1) tss else isValidMOf n tss
+    go = \case
+      RequireTimeStart lockStart -> lockStart `lteNegInfty` txStart
+      RequireTimeExpire lockExp -> txExp `ltePosInfty` lockExp
+      RequireSignature hash -> hash `Set.member` vhks
+      RequireAllOf xs -> all go xs
+      RequireAnyOf xs -> any go xs
+      RequireMOf m xs -> isValidMOf m xs
```

Comment:
could we call this something like 'validateTimeloxk'

Reply:
Same here. `evalTimelock` is the name of the function. `isValidMOf` is only applicable to the `RequireMOf`

Reply:
`validateTimelock` is the top level function right here that calls `evalTimelock`:

https://github.com/input-output-hk/cardano-ledger/blob/71bf6ae648fdead3bf37d5ba69d46f9a35d29dc2/eras/allegra/impl/src/Cardano/Ledger/Allegra/Tx.hs#L72-L77